### DISCUSSION TITLE
Concerns about authors of duplicate kata which are candidates for retirement

### DISCUSSION THREAD

**Discussion by hobovsky** [codewars/content-issues#68] [General]
To continue the discussion started in [Deduplicate - Count Digits of Factorial (#67)](https://github.com/codewars/content-issues/issues/67)

[FArekkusu wrote](https://github.com/codewars/content-issues/issues/67#issuecomment-1020326741):

> > I lean towards keeping 1. as... being actively maintained
>  
> That is until you raise an issue and KenKamau blocks you. (It's rather surprising to see that not only nobody cares how much of a headache dealing with him is, but people also consider this an "active maintenance".)
>  
> Anyway, all the existing versions and translations on the second kata are up-to-date in terms of using the test frameworks properly, they all have a good(?) set of fixed tests and random tests. I'd rather see that one survive, possibly unpublished and reranked as `6 kyu` if the ranking is a problem, than the first one.

[hobovsky wrote](https://github.com/codewars/content-issues/issues/67#issuecomment-1020357083):

> > That is until you raise an issue and KenKamau blocks you. (It's rather surprising to see that not only nobody cares how much of a headache dealing with him is, but people also consider this an "active maintenance".)
> 
> It's not "people", it's specifically me. Many users share your opinion and many users did not hesitate to let me know about their concerns on this topic. So to explain myself a bit:
> 
> - Don't get me wrong: I also do not like how some authors treat other users. Or how other users treat other users, for that matter. I think I am not totally blind and can see the problem here, and in other places.
> - Specifically about Ken, I _think _ most issues reported to him are actually handled, just after the reporter gets blocked. But maybe I am wrong.
> - Since some time we (Codewars community) actually have means to deal with potentially problematic cases. Kata can be forked easily, and forks with fixes can be approved by mods.
> - I do really believe that kata quality should be a priority, and if you say that the other kata has better tests, that's noted. This is exactly what these discussions are for.
> 
> Please don't hijack this thread with concerns related to the author - I think they are not significant for this particular issue, and if anyone wants to discuss this specific matter, we can start a new discussion.
> 
> The problem with reranking is that until now, we have no really good means tor this. The kata needs to go through beta and needs to be reapproved with a new rank, and this is another can of worms to deal with (resolving issues, collecting votes, making sure it gets approved with a proper rank, getting admins involved more than necessary - it all takes time and effort). I see keeping properly ranked kata as just easier, because this is the most difficult part to fix. Until we get some better tools for reranking, we need to account for related difficulties.
> 
> If your opinion is "keep 2., retire 1." that's fine, thanks for sharing it. Let's just focus on the quality of the kata which is to be kept, and avoid the possibility that we go for worse option just because we think it will be easier to introduce.

[FArekkusu wrote](https://github.com/codewars/content-issues/issues/67#issuecomment-1020426196):

> > Please don't hijack this thread with concerns related to the author
> 
> Excuse me? Whether you want it or not, deduplicating katas is closely related to the topic of dealing with the surviving kata(s) and their author(s) in the future, and IMO the question of whether it's better to have a kata with an active but **very** counter-productive author or have a kata with no active author at all is a valid and relevant concern.
> 
> > Since some time we (Codewars community) actually have means to deal with potentially problematic cases. Kata can be forked easily, and forks with fixes can be approved by mods.
> 
> If you're suggesting that we (Codewars community) can deal with any issues ourselves, without the author's involvement, and we should "just focus on the quality of the kata", then why do you bring up the author's status in the first place? To create a fake argument which doesn't really mean anything?

**Comment by hobovsky** [comment]
I wanted to create some background and  prepare a complete description of the proposal of deduplication process and how it's conducted now, but unfortunately i have not enough time :( That's why sometimes during the discussion I might introduce some ideas which were not presented or explained yet, but everything is still just ideas. Be aware that there's no exact definition of handling duplicates yet, and all threads and discussions are meant to get as many people engaged as possible, collect opinions, and build something on top of them.

> Excuse me? Whether you want it or not, deduplicating katas is closely related to the topic of dealing with the surviving kata(s) and their author(s) in the future, and IMO the question of whether it's better to have a kata with an active but very counter-productive author or have a kata with no active author at all is a valid and relevant concern.

I am not sure about this part, and actually I think I disagree. My (hobovsky's) idea is to keep authors responsible for maintenance of their kata, and do not allow them to obstruct any actions which would lead to increase of quality of the content. If an author, for some reason, neglects their kata, abandons it, or simply rejects to fix it, their decisions can be easily overriden by users with sufficient privileges (today these would be moderators). So I think that dealing with counter-productive authors goes (almost) completely out of picture, because if you think that a kata needs a fix and author does not want to introduce it, you can still prepare a fork and ask a mod to approve it. Authors do not have to be involved at all, and if their actions are somehow obstructive, they can be worked around. If you ask specifically me (hobovsky), in justified cases I would be ready to completely hijack a kata of a bad quality and introduce changes to it, using tools available to me, even when facing a visible objection of its author. I even have some candidate kata for this.

> If you're suggesting that we (Codewars community) can deal with any issues ourselves, without the author's involvement, and we should "just focus on the quality of the kata", then why do you bring up the author's status in the first place? To create a fake argument which doesn't really mean anything?

I include information about authors in a summary of duplicates because other reviewers/maintainers/mods sometimes ask about it. I do not know why, and I am not sure how it matters for them. For me  (hobovsky), it totally does not matter because, frankly, I do not really care about what authors have to say about their kata being a duplicate. Either it is recognized as one by community, or not - that's all. Please note how the bullet point about author appears as the last in the summary - I try to put the points in order of relevance (perceived by me).
The format of the summary is not established yet, I just put there whatever I consider meaningful for potential reviewers. If you think something is not necessary, or something is missing, you can always let me know.

As I said before, the whole process of removing duplicates, and even a definition of a duplicate, is not established yet. I just happened to have some more time and decided to pick some kata from the list, the most obvious ones first, and try something. Get as many people involved as possible, see how things go, what ideas pop up, and eventually improve. Feel free to discuss and criticize anything related to the topic, that's why I create all the discussions in the first place.

**Reply by mubii007** [reply]
Hi,
My profile link is [here](https://www.codewars.com/users/Mubashir%20Hassan). Can you please remove restrictions on my account? You can delete all solutions which you think are copied. :(

**Comment by ejini6969** [comment]
> That is until you raise an issue and KenKamau blocks you. (It's rather surprising to see that not only nobody cares how much of a headache dealing with him is, but people also consider this an "active maintenance".)

I think revoking of author's or any CW users' authority in blocking others can be done by admins if they continue to block those who give them constructive feedback (either in identifying existing kata's issues or suggesting language updates or what not). 

> Excuse me? Whether you want it or not, deduplicating katas is closely related to the topic of dealing with the surviving kata(s) and their author(s) in the future, and IMO the question of whether it's better to have a kata with an active but very counter-productive author or have a kata with no active author at all is a valid and relevant concern.

Yes, I agree, author plays an important criteria in deduplicating process. IMO, keeping the kata with an active author is a better choice since it is considered disrespectful to retire a kata when the author is still present (showing some form of responsibility in maintaining his/her own kata). However, counter-productive authors is not a problem too as we have initiatives to deal with them ⬇️⬇️⬇️

> My (hobovsky's) idea is to keep authors responsible for maintenance of their kata, and do not allow them to obstruct any actions which would lead to increase of quality of the content. If an author, for some reason, neglects their kata, abandons it, or simply rejects to fix it, their decisions can be easily overriden by users with sufficient privileges (today these would be moderators). So I think that dealing with counter-productive authors goes (almost) completely out of picture, because if you think that a kata needs a fix and author does not want to introduce it, you can still prepare a fork and ask a mod to approve it. Authors do not have to be involved at all, and if their actions are somehow obstructive, they can be worked around. If you ask specifically me (hobovsky), in justified cases I would be ready to completely hijack a kata of a bad quality and introduce changes to it, using tools available to me, even when facing a visible objection of its author. I even have some candidate kata for this.

~~ As said by Hobovsky above. 

> As I said before, the whole process of removing duplicates, and even a definition of a duplicate, is not established yet. I just happened to have some more time and decided to pick some kata from the list, the most obvious ones first, and try something. Get as many people involved as possible, see how things go, what ideas pop up, and eventually improve. Feel free to discuss and criticize anything related to the topic, that's why I create all the discussions in the first place.

Thanks @hobovsky for your efforts ! Just my opinion, I think everyone's opinion regarding duplicate katas still differ a lot, so more discussion is required. But for my opinion, I think we can establish few criterion such as 

i) satisfaction rating
ii) total completions
iii) Quality and validity of tests across all languages (including language updates)
iv) Task description (input range, validation required or not, user-friendliness)
v) NO. of pending translations
vi) ranking
etc.

Note that the above criterion do not have any significance order.

We can also need not to consider kata's creation data as a criteria during deduplication process.

**Reply by EloiseRosen** [reply]
> We can also need not to consider kata's creation data as a criteria during deduplication process.

I could see an argument for saying that the first kata wasn't a duplicate at the time it was made, but subsequent ones were, and that it's an author's responsibility to check if their idea has already been done.

**Comment by Kacarott** [comment]
It is of course important to prioritize kata quality when choosing between duplicates, however I think that considering the effort involved, (now and in the future) is also important. After all, any kata can be converted into any other kata, it is merely a matter of how much time and effort is spent on it. That is why we choose katas with more translations, that they need not be translated again, and why language agnostic descriptions are preferred.

And so, when considering between duplicates, I think the author *is* relevant to the discussion. An active, and helpful author is best case, as it hopefully means less effort from the community on fixing issues in the future. An active, but purposely unhelpful and difficult author, however, is actually worse than one that is completely inactive as they increase the effort required to do *anything* to improve the kata.

In Ken's case, even a task as simple as updating a language version (to Python 3.8) was quite a hassle, requiring multiple raised issues, and finally just making the modification myself once the issues were old enough. And of course, I was rewarded with a block.
